ARExperiment/ARExperiment/RecognizeObjectsViewController.swift

```diff
@@ -92,78 +87,172 @@ class RecognizeObjectsViewController: UIViewController, ARSCNViewDelegate, ARExp
     }
     
     private func setUpVision() {
-        guard let visionModel = try? VNCoreMLModel(for: yolo.model.model) else {
-            print("Error: could not create Vision model")
-            return
+        if usingTinyModel {
+            guard let visionModel = try? VNCoreMLModel(for: yolo.model.model) else {
+                print("Error: could not create Vision model")
+                return
+            }
+            request = VNCoreMLRequest(model: visionModel, completionHandler: visionRequestDidComplete)
+        } else {
+            guard #available(iOS 12.0, *) else {
+                usingTinyModel = true
+                setUpVision()
+                return
+            }
+            guard let visionModel = try? VNCoreMLModel(for: ObjectDetector().model) else {
+                print("Error: could not create Vision model")
+                return
+            }
+            request = VNCoreMLRequest(model: visionModel, completionHandler: visionRequestDidComplete)
         }
-        request = VNCoreMLRequest(model: visionModel, completionHandler: visionRequestDidComplete)
-        
+
         // NOTE: If you choose another crop/scale option, then you must also
         // change how the BoundingBox objects get scaled when they are drawn.
         // Currently they assume the full input image is used.
         request.imageCropAndScaleOption = .scaleFill
     }
     
     func visionRequestDidComplete(request: VNRequest, error: Error?) {
-        if let observations = request.results as? [VNCoreMLFeatureValueObservation],
-            let features = observations.first?.featureValue.multiArrayValue {
-            
-            let boundingBoxes = yolo.computeBoundingBoxes(features: features)
-            showOnMainThread(boundingBoxes)
+        self.semaphore.signal()
```

Comment:
oo why do we need a semaphore?!

Reply:
Because if not we get too many signals at once on the ML and it gets kinda confused a lot, I got really bad prediction results so when I was looking on how to work with machine learning and AR they recommended a semaphore that was basically only allowing one image from camera at a time to go through the model until the model is done. To know overrun the queque and make sure we get better results.

Reply:
OK, interesting! There might be a better way to do this but I'll have a think